### PR TITLE
[CALCITE-3348] AssertionError while determining distribution of Calc

### PR DESCRIPTION
**Pull request by yanlin-Lynn**
When using `EnumerableCalcRule` with `RelDistributionTraitDef` for planner, got AssertionError.
Please refer to [CALCITE-3348](https://issues.apache.org/jira/browse/CALCITE-3348) for details.

This PR tries to implement the logic of determining distribution for Calc.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistribution.java [hunk 1]
```diff
@@ -141,10 +141,17 @@ public static RelDistribution limit(RelMetadataQuery mq, RelNode input) {
    * {@link org.apache.calcite.rel.core.Calc}'s distribution. */
   public static RelDistribution calc(RelMetadataQuery mq, RelNode input,
       RexProgram program) {
-    throw new AssertionError(); // TODO:
+    assert program.getCondition() != null || !program.getProjectList().isEmpty();
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistribution.java]: What's the assertion for ?
It seems common case that program.getCondition is null when there's no filtering condition.

File: core/src/main/java/org/apache/calcite/rex/RexProgram.java [hunk 2]
```diff
@@ -821,6 +824,29 @@ public RexProgram normalize(RexBuilder rexBuilder, boolean simplify) {
...
+   *
+   * @param inputFieldCount Number of input fields
+   * @return Mapping of a set of project expressions, never null
+   */
+  public Mappings.TargetMapping getPartialMapping(int inputFieldCount) {
+    Mappings.TargetMapping mapping =
+        Mappings.create(MappingType.INVERSE_FUNCTION,
+            inputFieldCount, projects.size());
+    for (Ord<RexLocalRef> exp : Ord.zip(projects)) {
+      RexInputRef inputRef = (RexInputRef) exprs.get(exp.e.index);
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/rex/RexProgram.java]: Shall we call `RexProgram#expandLocalRef` for the expansion ?

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistribution.java (hunk 1 above)
Comment (yanlin-Lynn) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdDistribution.java]: Make sure that the program is valid.
As my understanding, empty project list and null condition cannot come together.

File: core/src/main/java/org/apache/calcite/rex/RexProgram.java (hunk 2 above)
Comment (yanlin-Lynn) [reply core/src/main/java/org/apache/calcite/rex/RexProgram.java]: Yes, that's better.
Thanks for remindering me.

**Pull request merged**